rfd/0019-event-iteration-api.md

```diff
@@ -0,0 +1,110 @@
+---
+authors: Joel Wejdenstal (jwejdenstal@goteleport.com)
+state: draft
+---
+
+# RFD 19 - Event Fetch API with Pagination
+
+## What 1
+
+Implement a new API based on the current API for fetching events but with full pagination support in form of a `StartKey: string` parameter that allows the client to search from a specific point in the event stream instead of starting at the first event of the start date.
+
+## Why 2
+
+This RFC was sparked from issue [#5435](https://github.com/gravitational/teleport/issues/5435).
+The primary motivation for this RFD is that clients are at the moment not able to fetch an event stream between two dates incrementally with multiple API calls.
+
+Such functionality requires the client to be able to specify a point in the event stream from where the server will start searching forward from. Currently no such parameter exists which means the server will always start returning events from the start of the stream.
+
+## Details 3
+
+The current API for event fetch and search is over (https://github.com/gravitational/teleport/blob/master/lib/auth/apiserver.go#L246) plain HTTPS and not GRPC and runs on the authentication server.
+
+I propose implementing a new GRPC API that replaces and deprecates the old HTTP API endpoints. The old API endpoints should be kept as is but documentation should mention that it is deprecated and advise clients to use the new API.
+
+### GRPC Messages 3.1
+
+Out-of-session event request.
+This replaces and deprecates the `GET /:version/events` HTTP endpoint.
+
+These should be defined in [api/client/proto/authservice.proto](https://github.com/gravitational/teleport/blob/master/api/client/proto/authservice.proto).
+
+```protobuf
+message GetEventsRequest {
+   // Namespace, if not set, defaults to 'default'
+   Namespace string
+   // Oldest date of returned events
+   StartDate Timestamp
+   // Newest date of returned events
+   EndDate Timestamp
+   // EventType is optional, if not set, returns all events
+   EventType string
+   // Maximum amount of events returned
+   Limit int64
+   // When supplied the search will resume from the last key
+   StartKey string
+}
+```
+
+In-session event request.
+This replaces and deprecates the `GET /:version/events` HTTP endpoint.
+
+```protobuf
+message GetSessionEventsRequest {
+   // SessionID is a required valid session ID
+   SessionID string
+   EventType string
+   Limit int64
+   StartKey string
+}
+```
+
+Event response from both endpoints.
+
+```protobuf
+message Events { 
+   Items repeated oneof Event
+   // the key of the last event if the returned set did not contain all events found i.e limit < actual amount. this is the key clients can supply in another API request to continue fetching events from the previous last position
+   LastKey string
```

Comment:
I wonder if we can include a `total` field to the response?  The support for it will vary between data sources but we can return `-1` when not supported.

Reply:
Technically we can do it but for pretty much all backends it would slow down queries a fair bit. In my opinion we can have the client perform this calculation if needed.

Reply:
What do you mean by client performing these calculations?

Reply:
Technically we can implement everything:) but realistically we should be clear that total amount will not be supported then. Which is fine if everyone agrees.

Reply:
I'm assuming `total` here is the amount of events returned by this message and all events in the date selection preceding. Figuring out that number means that we have to at least select and count all events preceding the `StartKey` associated with this request which significantly removes performance benefits that we get by pagination since we would rescan events over and over again.

Reply:
I.E unless I am missing something, calculating `total` would place `O(n)` stress on the datastore.

Reply:
If client repeatedly fetches events, it should be able to calculate total by summing the all amounts of events it has received previously assuming `StartKey` is always the previous `LastKey`.

Reply:
There 2 things:
1. Scanning the events, running actual query in the DB.
2. Returning the results which may be gigabytes of data. 

Both of them addressed differently. In here we are mostly focusing on #2 as this is the major bottleneck at the moment. Running an actual query on pre-processed by backend data might be faster then we think but I guess it's out of scope.

Reply:
Running the actual query in the DB isn't as bad as returning data but it can cause an explosion in load on very busy days since every request will have to rescan events on the backend as is currently without some form of additional state tracking. This is a solvable problem but I think that addressing it is out of scope.

Reply:
Based on my read it does sound like a to solve later problem if we look at ES or Splunk etc type backends. The way splunk works for example is data is pulled to a search head to keep state and the client keeps a cursor into the data. There search also works on a streaming concept, so in the UI you can start to interact with results while it's still building the results if doing a really large search.

It's a really great system, but seems way more complicated than this design is attempting to target.